cmd/nitro/init.go

```diff
@@ -811,178 +818,332 @@ func openInitializeExecutionDB(ctx context.Context, stack *node.Node, config *No
+		} else {
+			log.Warn("failed to open l2chaindata", "err", err)
+		}
+	}
+
+	return nil, nil, nil, nil, nil
+}
+
+func getConsensusParsedInitMsg(ctx context.Context, config *NodeConfig, chainId *big.Int, l1Client *ethclient.Client, rollupAddrs chaininfo.RollupAddresses, chainConfig *params.ChainConfig) (*arbostypes.ParsedInitMessage, error) {
+	var parsedInitMessage *arbostypes.ParsedInitMessage
+	if config.Node.ParentChainReader.Enable {
+		delayedBridge, err := arbnode.NewDelayedBridge(l1Client, rollupAddrs.Bridge, rollupAddrs.DeployedAt)
+		if err != nil {
+			return nil, fmt.Errorf("failed creating delayed bridge while attempting to get serialized chain config from init message: %w", err)
+		}
+		deployedAt := new(big.Int).SetUint64(rollupAddrs.DeployedAt)
+		delayedMessages, err := delayedBridge.LookupMessagesInRange(ctx, deployedAt, deployedAt, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed getting delayed messages while attempting to get serialized chain config from init message: %w", err)
+		}
+		var initMessage *arbostypes.L1IncomingMessage
+		for _, msg := range delayedMessages {
+			if msg.Message.Header.Kind == arbostypes.L1MessageType_Initialize {
+				initMessage = msg.Message
+				break
+			}
+		}
+		if initMessage == nil {
+			return nil, fmt.Errorf("failed to get init message while attempting to get serialized chain config")
+		}
+		parsedInitMessage, err = initMessage.ParseInitMessage()
+		if err != nil {
+			return nil, err
+		}
+		if parsedInitMessage.ChainId.Cmp(chainId) != 0 {
+			return nil, fmt.Errorf("expected L2 chain ID %v but read L2 chain ID %v from init message in L1 inbox", chainId, parsedInitMessage.ChainId)
+		}
+		if parsedInitMessage.ChainConfig != nil {
+			if err := parsedInitMessage.ChainConfig.CheckCompatible(chainConfig, chainConfig.ArbitrumChainParams.GenesisBlockNum, 0); err != nil {
+				return nil, fmt.Errorf("incompatible chain config read from init message in L1 inbox: %w", err)
+			}
+		}
+		log.Info("Read serialized chain config from init message", "json", string(parsedInitMessage.SerializedChainConfig))
+	} else {
+		serializedChainConfig, err := json.Marshal(chainConfig)
+		if err != nil {
+			return nil, err
+		}
+		parsedInitMessage = &arbostypes.ParsedInitMessage{
+			ChainId:               chainConfig.ChainID,
+			InitialL1BaseFee:      arbostypes.DefaultInitialL1BaseFee,
+			ChainConfig:           chainConfig,
+			SerializedChainConfig: serializedChainConfig,
+		}
+		log.Warn("Created fake init message as L1Reader is disabled and serialized chain config from init message is not available", "json", string(serializedChainConfig))
+	}
+
+	return parsedInitMessage, nil
+}
+
+func getGenesisAssertionCreationInfo(ctx context.Context, initDataReader *statetransfer.InitDataReader, rollupAddress common.Address, l1Client *ethclient.Client, genesis *types.Block) (*protocol.AssertionCreatedInfo, [32]byte, bool, error) {
```

Comment:
In a future PR, getGenesisAssertionCreationInfo will only be called by the Consensus component.
The Consensus component will not have access to to statetransfer.InitDataReader, this is something that only Execution will have access to.
Execution will retrieve from Execution genesis's block hash and send root through a Execution.ResultAtMessageIndex call, to later call getGenesisAssertionInfo.

This func signature should be something like this:
`func getGenesisAssertionInfo(ctx context.Context, rollupAddress common.Address, l1Client *ethclient.Client, genesisHash common.Hash, genesisSendRoot common.Hash) (genesisAssertionHash [32]byte, genesisAssertionCreationInfo *protocol.AssertionCreatedInfo, error)`

Reply:
InitDataReader is a property from Execution.
Later, getGenesisAssertionInfo will only be executed by Consensus, and Consensus doesn't have access to InitDataReader, nor initDataReaderHasAccounts.
validateGenesisAssertion will only be executed by Execution, and Execution has access to InitDataReader and then initDataReaderHasAccounts.
Therefore initDataReaderHasAccounts computation can be inside validateGenesisAssertion, or validateGenesisAssertion can receive it as an argument.
It is OK to have this getAndValidateGenesisAssertion func right now :), but it will be deleted after validateGenesisAssertion is moved to ExecutionNode in a future PR